internal/evaluator/evaluator.go

```diff
@@ -0,0 +1,344 @@
+func evalIfExpression(ie *ast.IfExpression, env *object.Environment) object.Object {
+	condition := Eval(ie.TestCondition, env)
+	if isError(condition) {
+		return condition
+	}
+	if isTruthy(condition) {
+		return Eval(ie.ThenPath, env)
+	} else if ie.ElsePath != nil {
+		return Eval(ie.ElsePath, env)
+	} else {
+		return NULL
+	}
+}
+
+// Truthy means that values other than False and null are evaluate to be true
+func isTruthy(obj object.Object) bool {
+	switch obj {
+	case NULL:
+		return false
+	case TRUE:
+		return true
+	case FALSE:
+		return false
+	default:
+		return true
+	}
+}
+
+// Recursively evaluates a series of program statements then returns the return value,
+// else the last result
+// Stops evaluating early if there is a return statement
+func evalProgram(program *ast.Program, env *object.Environment) object.Object {
+	var result object.Object
+	for _, statement := range program.Statements {
+		result = Eval(statement, env)
+		switch result := result.(type) {
+		case *object.ReturnValue:
+			return result.Value
+		case *object.Error:
+			return result
+		}
+	}
+	return result
+}
+
+// Recursively evaluates a series of block statements then returns the result
+func evalBlockStatement(block *ast.BlockStatement, env *object.Environment) object.Object {
+	var result object.Object
+	for _, statement := range block.Statements {
+		result = Eval(statement, env)
+		if result != nil {
+			rt := result.Type()
+			if rt == object.RETURN_VALUE_OBJ || rt == object.ERROR_OBJ {
+				return result
+			}
+		}
+	}
+	return result
+}
+
+// Returns true if the object is an error object, else false
+func isError(obj object.Object) bool {
+	if obj != nil {
+		return obj.Type() == object.ERROR_OBJ
+	}
+	return false
+}
+
+// Returns the value of an identifier stored in an enviornment
+func evalIdentifier(
+	node *ast.Identifier,
+	env *object.Environment,
+) object.Object {
+	val, ok := env.Get(node.Value)
+	if !ok {
+		return newError("identifier not found: " + node.Value)
+	}
+	return val
+}
+
+// Evaluates a list of expressions
+// Gets the result of each evaluated expression and stores in a list as a tuple,
+// with a list of results from expressions evaluated before
+func evalExpressions(
+	exps []ast.Expression,
+	env *object.Environment,
+) []object.Object {
+	var result []object.Object
+	for _, e := range exps {
+		evaluated := Eval(e, env)
+		if isError(evaluated) {
+			return []object.Object{evaluated}
+		}
+		result = append(result, evaluated)
+	}
+	return result
+}
+
+// Checks that we have a function
+// Creates a new environment with function object and args,
+// then evaluates the function body.
+// Returns result of the function call
+func applyFunction(fn object.Object, args []object.Object) object.Object {
+	function, ok := fn.(*object.Function)
+	if !ok {
+		return newError("not a function: %s", fn.Type())
+	}
+	extendedEnv := extendFunctionEnv(function, args)
```

Comment:
To fix the wrong arity issue aka not enough arguments being passed into functions.

```go
        if len(function.Parameters) != len(args) {
                return newError(
                        "wrong number of arguments. want=%d. got=%d",
                        len(function.Parameters),
                        len(args),
                )
        }
```

Reply:
Make sure to add a test for this case. One for too little, and too many.